### PR TITLE
Refactor: Separate CKeyStore interface from its basic implementation

### PR DESCRIPTION
**Pull request by jtimon**
And move CCriticalSection cs_KeyStore up from CKeyStore to CBasicKeyStore. 
Also cleanup some dependencies.
This serves as preparation for #5208

### CODE REVIEW

File: src/keystore.h [hunk 1]
```diff
@@ -19,21 +15,28 @@ class CScriptID;
...
+    {
+        CKey key;
+        if (!GetKey(address, key))
+            return false;
+        vchPubKeyOut = key.GetPubKey();
+        return true;
+    }
 
     //! Support for BIP 0013 : see https://github.com/bitcoin/bips/blob/master/bip-0013.mediawiki
     virtual bool AddCScript(const CScript& redeemScript) =0;
```
Comment (laanwj) [inline src/keystore.h]: If the new keystore is meant to be a simple private key store for e.g. raw transaction API usage, then do `*WatchOnly` methods still belong on it?

Comment (jtimon) [inline src/keystore.h]: Sorry for not answering earlier. If you comment those methods you'll get this errors:
```
wallet_ismine.cpp: In function ‘isminetype IsMine(const CKeyStore&, const CScript&)’:
wallet_ismine.cpp:42:22: error: ‘const class CKeyStore’ has no member named ‘HaveWatchOnly’
         if (keystore.HaveWatchOnly(scriptPubKey))
                      ^
wallet_ismine.cpp:88:18: error: ‘const class CKeyStore’ has no member named ‘HaveWatchOnly’
     if (keystore.HaveWatchOnly(scriptPubKey))
```
The other 3 methods can be safely removed from the interface.
I mean, it's not unsolvable but I wanted to keep this simple...

**Comment by jtimon** [comment]
Needs rebase but @theuni was working on a better alternative. Should I rebase it or not?

**Comment by theuni** [comment]
I have something but I'm not sure if it's worth using. I'm away until
Monday so don't hold on me if this blocks further work.

**Comment by sipa** [comment]
I think CBasicKeyStore can actually just go in script/sign; it's basically the interface that users of script/sign need to implement to enable arbitrary-key signing.

**Comment by jtimon** [comment]
@sipa that makes sense to me. Anyway, it already needs rebase and I would probably close this in favor of @theuni 's improved proposal once it's ready, so I'm closing for now.

**Pull request closed without merging**